### PR TITLE
[test] moe: add a more thorough MOELayer routing test

### PR DESCRIPTION
**Pull request by msbaines**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Fixes # (issue).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/nn/moe/test_moe_layer.py [hunk 1]
```diff
@@ -73,6 +73,56 @@ def test_forward(device):
...
+# Test Gate which round-robin routes tokens to experts
+class RoundRobinGate(torch.nn.Module):
+    def __init__(self, model_dim, num_experts):
+        super().__init__()
+        self.model_dim = model_dim
+        self.num_experts = num_experts
+
+    def forward(self, input):
+        g, s, m = input.shape
+        capacity = 2 * s // self.num_experts
```
Comment (min-xu-ai) [inline tests/nn/moe/test_moe_layer.py:85]: it seems that "//" should be "/" because if it is not divided evenly, then you should get an error anyway. "//" will mask error?

File: tests/nn/moe/test_moe_layer.py [hunk 2]
```diff
@@ -73,6 +73,56 @@ def test_forward(device):
...
+        output = torch.zeros(g, s, self.num_experts, capacity, dtype=input.dtype, device=input.device)
```
Comment (min-xu-ai) [inline tests/nn/moe/test_moe_layer.py:86]: I wonder if you can do something like:

```
o = []
for I in range(self.num_experts):
  x = np.array((g, s, 1, capacity))
  x[:,:,:,self.num_experts*I:self.num_experts*(I+1)] = 1.0
  o.append(x)
o = np.stack(o)
output = torch.tensor(o)
```

I didn't write the exact np code above, but just rough idea. That would avoid explicitly looping over all indexes and the meaning is more readable too?

**Review by min-xu-ai — COMMENTED** [review]
some comments.

File: tests/nn/moe/test_moe_layer.py (hunk 1 above)
Comment (msbaines) [reply tests/nn/moe/test_moe_layer.py:86]: Added an assertion.

**Review by msbaines — COMMENTED** [review]
Fixed. PTAL.

File: tests/nn/moe/test_moe_layer.py [hunk 3]
```diff
@@ -73,6 +73,56 @@ def test_forward(device):
...
+
+    def forward(self, input):
+        g, s, _ = input.shape
+        assert s % self.num_experts == 0
+        capacity = 2 * s // self.num_experts
+        output = torch.zeros(g, s, self.num_experts, capacity, dtype=input.dtype, device=input.device)
+        expert = 0
+        c = 0
+        for i in range(s):
+            output[:, i, expert, c] = 1.0
```
Comment (min-xu-ai) [inline tests/nn/moe/test_moe_layer.py:91]: it seems `expert = i % self.num_experts` and `c = i // self.num_experts`? No need to have separate variables and ++ them?

File: tests/nn/moe/test_moe_layer.py [hunk 4]
```diff
@@ -73,6 +73,56 @@ def test_forward(device):
...
+    gate = RoundRobinGate(model_dim, num_experts)
+    expert = torch.nn.Linear(model_dim, model_dim, bias=False)
+    # Use scaling matrix (each rank has a different scale)
+    scale = dist.get_rank() + 1
+    expert.weight = torch.nn.Parameter(torch.eye(model_dim) * scale)
+    moe = MOELayer(gate, expert).to(device)
+    output = moe(input)
+    assert output.shape == input.shape
+    # Verify that each token was sent to the correct expert by checking its scale.
+    g, s, t, m = input.shape
```
Comment (min-xu-ai) [inline tests/nn/moe/test_moe_layer.py:115]: same here, m can be replaced with _.

File: tests/nn/moe/test_moe_layer.py [hunk 5]
```diff
@@ -73,6 +73,56 @@ def test_forward(device):
...
+    expert = 0
+    for i in range(s):
+        for j in range(t):
+            for k in range(g):
+                torch.allclose(input[k][i][j], output[k][i][j] * (expert + 1))
```
Comment (min-xu-ai) [inline tests/nn/moe/test_moe_layer.py:120]: expert = (i * s + j * t)//num_experts?

File: tests/nn/moe/test_moe_layer.py [hunk 6]
```diff
@@ -73,6 +73,56 @@ def test_forward(device):
...
+    expert.weight = torch.nn.Parameter(torch.eye(model_dim) * scale)
+    moe = MOELayer(gate, expert).to(device)
+    output = moe(input)
+    assert output.shape == input.shape
+    # Verify that each token was sent to the correct expert by checking its scale.
+    g, s, t, m = input.shape
+    expert = 0
+    for i in range(s):
+        for j in range(t):
+            for k in range(g):
```
Comment (min-xu-ai) [inline tests/nn/moe/test_moe_layer.py:119]: similarly perhaps this loop can be avoided by use ":" indexing?

**Review by min-xu-ai — APPROVED** [review]
LGTM. minor suggestions to make it consistent with your latest change.
